vowpalwabbit/core/src/reductions/eigen_memory_tree.cc

```diff
@@ -0,0 +1,1019 @@
+// Copyright (c) by respective owners including Yahoo!, Microsoft, and
+// individual contributors. All rights reserved. Released under a BSD (revised)
+// license as described in the file LICENSE.
+
+#include "vw/core/reductions/eigen_memory_tree.h"
+
+#include "vw/common/future_compat.h"
+#include "vw/config/options.h"
+#include "vw/core/array_parameters.h"
+#include "vw/core/example.h"
+#include "vw/core/learner.h"
+#include "vw/core/numeric_casts.h"
+#include "vw/core/rand48.h"
+#include "vw/core/rand_state.h"
+#include "vw/core/setup_base.h"
+#include "vw/core/v_array.h"
+#include "vw/core/vw.h"
+#include "vw/io/logger.h"
+
+#include <algorithm>
+#include <cfloat>
+#include <climits>
+#include <cmath>
+#include <cstdio>
+#include <ctime>
+#include <list>
+#include <memory>
+#include <sstream>
+
+using namespace VW::LEARNER;
+using namespace VW::config;
+
+// TODO: This file has several cout print statements. It looks like
+//       they should be using trace_message, but its difficult to tell
+namespace
+{
+///////////////////////Helper//////////////////////////////
+//////////////////////////////////////////////////////////
+float median(std::vector<float>& array)
+{
+  int size = array.size();
+  std::sort(array.begin(), array.end());
+
+  if (size % 2 == 0) { return (array[size / 2 - 1] + array[size / 2]) / 2; }
+  else
+  {
+    return array[size / 2];
+  }
+}
+
+float variance(std::vector<float>& array)
+{
+  float E = std::accumulate(array.begin(), array.end(), 0.0) / array.size();
+  float E_2 = std::inner_product(array.begin(), array.end(), array.begin(), 0.0) / array.size();
+  return E_2 - E * E;
+}
+
+/// <summary>
+/// DO NOT USE abs() UNLESS YOU WANT TO SPEND A DAY DEBUGGING WEIRD BUGS ON REMOTE SERVERS
```

Comment:
I don't really know much about c++ so there's probably an obvious solution but it seems to be a problem affecting other people as well (https://stackoverflow.com/q/21392627/1066291). I'm happy to fix it if there's a better way rather than writing a silly custom abs function.